lib/explorer/shared.ex

```diff
@@ -281,6 +280,8 @@ defmodule Explorer.Shared do
         end
       end)
 
+    # Logger.info("type - #{inspect type}")
+    # || preferable_type || {:f, 64}
     type || preferable_type || {:f, 64}
```

Comment:
I think we can instead change this line to:

```suggestion
    type || preferable_type || :null
```

And then we can most likely skip other changes to this file?

Reply:
To be precise, skip the changes to `type/2` and `new_type_matches?`. :)

Reply:
3 existing test cases fail :
```elixir
s = Series.from_list([])
p = Series.cumulative_product(s)
**  (ArgumentError) Explorer.Series.cumulative_product/2 not implemented for dtype :null.
p = Explorer.Series.product(s)
**  (ArgumentError) Explorer.Series.product/1 not implemented for dtype :null
p = Series.less(s, nil)
**  cannot invoke Explorer.Series.less/2 with mismatched dtypes: {:f, 64} and nil

```
2 new test cases fail:
```elixir
s = Series.from_list([1, 2, 3], dtype: :null)
** (ArgumentError) the value 1 does not match the inferred series dtype :null
s =
     Series.from_list([
       %{a: nil, b: nil},
       %{a: 3, b: nil},
       %{a: 5, b: nil}
     ])
** (CaseClauseError) no case clause matching: nil
```

There is assumption that empty array is `{:f, 64}`